tests/test_integrated_channels/test_canvas/test_client.py

```diff
@@ -86,3 +89,144 @@ def test_client_instantiation_fails_without_refresh_token(self):
+                    json={'access_token': self.access_token},
+                    status=200
+                )
+                transmitter_method = getattr(canvas_api_client, request_type)
+                transmitter_method(empty_data)
+
+        assert client_error.value.__str__() == 'Canvas Client Error: No data to transmit.'
+
+    def test_course_delete_fails_with_poorly_formatted_data(self):
+        self.update_fails_with_poorly_formatted_data("_delete")
+
+    def test_course_update_fails_with_poorly_formatted_data(self):
+        self.update_fails_with_poorly_formatted_data("_put")
+
+    def update_fails_with_poorly_formatted_data(self, request_type):
+        poorly_formatted_data = 'this is a string, not a bytearray'
+        canvas_api_client = CanvasAPIClient(self.enterprise_config)
+
+        with pytest.raises(CanvasClientError) as client_error:
+            with responses.RequestsMock() as request_mock:
+                request_mock.add(
+                    responses.POST,
+                    self.oauth_url,
+                    json={'access_token': self.access_token},
+                    status=200
+                )
+                transmitter_method = getattr(canvas_api_client, request_type)
+                transmitter_method(poorly_formatted_data)
+
+        assert client_error.value.__str__() == 'Canvas Client Error: Unable to decode data.'
+
+    def test_course_delete_fails_with_poorly_constructed_data(self):
+        self.update_fails_with_poorly_constructed_data("_delete")
+
+    def test_course_update_fails_with_poorly_constructed_data(self):
+        self.update_fails_with_poorly_constructed_data("_put")
+
+    def update_fails_with_poorly_constructed_data(self, request_type):
+        bad_course_to_update = '{"course": {"name": "test_course"}}'.encode()
+        canvas_api_client = CanvasAPIClient(self.enterprise_config)
+
+        with pytest.raises(CanvasClientError) as client_error:
+            with responses.RequestsMock() as request_mock:
+                request_mock.add(
+                    responses.POST,
+                    self.oauth_url,
+                    json={'access_token': self.access_token},
+                    status=200
+                )
+                transmitter_method = getattr(canvas_api_client, request_type)
+                transmitter_method(bad_course_to_update)
+
+        assert client_error.value.__str__() == 'Canvas Client Error: ' \
+                                               'Could not transmit data, no integration ID present.'
+
+    def test_course_delete_fails_when_course_id_not_found(self):
+        self.update_fails_when_course_id_not_found("_delete")
+
+    def test_course_update_fails_when_course_id_not_found(self):
+        self.update_fails_when_course_id_not_found("_put")
+
+    def update_fails_when_course_id_not_found(self, request_type):
+        course_to_update = '{{"course": {{"integration_id": "{}", "name": "test_course"}}}}'.format(
```

Comment:
this is just me but is there a way to make this json formatting in the code clearner? no biggie but if you have a quick way to do so, multiline string?

Reply:
I can clean it up a bit 👍  it's kinda messy